couchbase/src/main/scala/akka/stream/alpakka/couchbase/scaladsl/CouchbaseCollectionSession.scala

```diff
@@ -0,0 +1,210 @@
+import akka.{Done, NotUsed}
+import com.couchbase.client.java.json.{JsonArray, JsonObject, JsonValue}
+import com.couchbase.client.java.kv.{InsertOptions, RemoveOptions, ReplaceOptions, UpsertOptions}
+import com.couchbase.client.java.manager.query.{CreateQueryIndexOptions, QueryIndex}
+import com.couchbase.client.java.{AsyncCollection, AsyncScope}
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+object CouchbaseCollectionSession {
+  protected def apply(bucket: CouchbaseSession, scope: String, collection: String) =
+    new CouchbaseCollectionSessionImpl(bucket, scope, collection)
+}
+
+@DoNotInherit
+trait CouchbaseCollectionSession {
+  def bucket: CouchbaseSession
+  def scope: AsyncScope
+  def underlying: AsyncCollection
+  def asJava: akka.stream.alpakka.couchbase.javadsl.CouchbaseCollectionSession
+
+  /**
+   * Insert a document using default InsertOptions and automatically detected Transcoder.
+   * Will use RawBinaryTranscoder if document._2 is an array of bytes, RawStringTranscoder if its a String
+   * and the default Transcoder otherwise.
+   * @param id id of the document
+   * @param document value of the document
+   * @tparam T type of the document
+   * @return a future that completes when the document was written
+   */
+  def insert[T](id: String, document: T): Future[Done]
+
+  /**
+   * Insert a document using provided InsertOptions.
+   * @param id id of the  document
+   * @param document value of the document
+   * @param insertOptions Couchbase InsertOptions
+   * @tparam T type of the document
+   * @return a future that completes when the document was written or errors if the operation failed
+   */
+  def insert[T](id: String, document: T, insertOptions: InsertOptions): Future[Done]
+
+  /**
+   * Reads a document with given id as JsonObject
+   * @param id id of the document
+   * @return a future that completes with the requested document or errors if the operation failed
+   */
+  def getJsonObject(id: String): Future[CouchbaseDocument[JsonObject]] =
+    get(id, classOf[JsonObject])
+
+  /**
+   * Reads a document with given id as JsonArray
+   * @param id id of the document
+   * @return a future that completes with the requested document or errors if the operation failed
+   */
+  def getJsonArray(id: String): Future[CouchbaseDocument[JsonArray]] = {
+    get(id, classOf[JsonArray])
+  }
+
+  /**
+   * Reads a document into an object of target class using appropriate Transcoders:
+   *  * If T === Array[Byte] then RawBinaryTranscoder is used
+   *  * If T === String then RawStringTranscoder is used
+   *  * The default json to class transcoder is used otherwise
+   * @param id id of the document
+   * @param target target class
+   * @tparam T type of the object to return
+   * @return a future that completes with created object or errors if the operation failed
+   */
+  def get[T](id: String, target: Class[T]): Future[CouchbaseDocument[T]]
```

Comment:
In Scala it would be more idiomatic to use:
```
def get[T: ClassTag](id: String): Future[CouchbaseDocument[T]]
```

If you need the actual class of that ClassTag you can use:
```
val clazz = implicitly[ClassTag[T]].runtimeClass
```

Reply:
I went ahead and applied this suggestion everywhere I was using generics, please lmk if that's not what you meant.